cmd/kubeadm/app/cmd/upgrade/apply.go

```diff
@@ -0,0 +1,192 @@
+	cmd.Flags().BoolVarP(&flags.force, "force", "f", flags.force, "Force upgrading although some requirements might not be met. This also implies non-interactive mode.")
+	cmd.Flags().BoolVar(&flags.dryRun, "dry-run", flags.dryRun, "Do not change any state, just output what actions would be applied. This also implies non-interactive mode.")
+	cmd.Flags().StringVar(&flags.newK8sVersionStr, "version", flags.newK8sVersionStr, "Selected version to upgrade to.")
+	cmd.Flags().DurationVar(&flags.pullTimeout, "pull-timeout", flags.pullTimeout, "The maximum amount of time to wait for the control plane pods to be downloaded.")
+
+	return cmd
+}
+
+// RunApply takes care of the actual upgrade functionality
+func RunApply(flags *applyFlags) error {
+
+	// Start with the basics, verify that the cluster is healthy and get the configuration from the cluster (using the ConfigMap)
+	upgradeVars, err := EnforceRequirements(flags.parent.kubeConfigPath, flags.parent.cfgPath, flags.parent.printConfig, flags.nonInteractiveMode)
+	if err != nil {
+		return err
+	}
+
+	// Set the upgraded version on the external config object now
+	upgradeVars.cfg.KubernetesVersion = flags.newK8sVersionStr
+
+	// Grab the external, versioned configuration and convert it to the internal type for usage here later
+	internalcfg := &kubeadmapi.MasterConfiguration{}
+	api.Scheme.Convert(upgradeVars.cfg, internalcfg, nil)
+
+	// Enforce the version skew policies
+	if err := EnforceVersionPolicies(flags, upgradeVars.versionGetter); err != nil {
+		return fmt.Errorf("[upgrade/version] FATAL: %v", err)
+	}
+
+	// If the current session is interactive, ask the user whether they really want to upgrade
+	if flags.SessionIsInteractive() {
+		if err := upgrade.InteractivelyConfirmUpgrade("Are you sure you want to proceed with the upgrade"); err != nil {
+			return err
+		}
+	}
+
+	// TODO: Implement a prepulling mechanism here
+
+	// Now; perform the upgrade procedure
+	if err := UpgradeControlPlaneComponents(flags, upgradeVars.client, internalcfg); err != nil {
+		return err
+	}
+
+	// Upgrade RBAC rules and addons. Optionally, if needed, perform some extra task for a specific version
+	if err := upgrade.PerformPostUpgradeTasks(upgradeVars.client, internalcfg, flags.newK8sVersion); err != nil {
+		return err
+	}
+
+	fmt.Println("")
+	fmt.Printf("[upgrade/successful] SUCCESS! Your cluster was upgraded to %q. Enjoy!\n", flags.newK8sVersionStr)
+	fmt.Println("")
+	fmt.Println("[upgrade/kubelet] Now that your control plane is upgraded, please proceed with upgrading your kubelets in turn.")
+
+	return nil
+}
+
+// ValidateFlags makes sure the arguments given to apply are valid
+func ValidateFlags(flags *applyFlags) error {
+	if len(flags.newK8sVersionStr) == 0 {
+		return fmt.Errorf("--version is a required parameter. Please specify the version you want to upgrade to")
+	}
+	var err error
```

Comment:
it was because of the assignment directly to cfg; now being explicit and using a temporary variable